src/main_context_channel.rs

```diff
@@ -18,74 +18,269 @@ use SourceId;
 use get_thread_id;
 
 use ffi as glib_ffi;
-use translate::{from_glib_full, mut_override, ToGlib, ToGlibPtr};
+use translate::{mut_override, FromGlibPtrFull, FromGlibPtrNone, ToGlib, ToGlibPtr};
+
+#[derive(Debug)]
+struct ChannelInner<T> {
+    queue: VecDeque<T>,
+    // This will be None when the receiver disappears and Some(NULL)
+    // before the receiver is attached
+    source: Option<*mut glib_ffi::GSource>,
+}
+
+unsafe impl<T> Send for ChannelInner<T> {}
+unsafe impl<T> Sync for ChannelInner<T> {}
+
+impl<T> ChannelInner<T> {
+    fn receiver_disconnected(&self) -> bool {
+        match self.source {
+            // Disconnected
+            None => true,
+            // Not attached or destroyed yet
+            Some(source) if source.is_null() => false,
+            // Receiver still there is destroyed already
+            Some(source)
+                if unsafe { glib_ffi::g_source_is_destroyed(source) } != glib_ffi::GFALSE =>
+            {
+                true
+            }
+            // Receiver still running
+            Some(_) => false,
+        }
+    }
+
+    fn set_ready_time(&mut self, ready_time: i64) {
+        if let Some(source) = self.source {
+            if !source.is_null() {
+                unsafe {
+                    glib_ffi::g_source_set_ready_time(source, ready_time);
```

Comment:
Maybe this function can return `true` if it really set time?

Reply:
It could but nothing would make use of that boolean :)

Reply:
Agree after use check